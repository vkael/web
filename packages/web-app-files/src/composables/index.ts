export * from './fileListHeaderPosition'
export * from './pagination'
export * from './resourcesViewDefaults'
export * from './router'
export * from './selection'
export * from './shares'
export * from './sort'
export * from './viewMode'
