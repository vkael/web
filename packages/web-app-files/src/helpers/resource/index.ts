export * from './ancestorMetaData'
export * from './filter'
export * from './sameResource'
export * from './conflictHandling'
export * from './actions'
