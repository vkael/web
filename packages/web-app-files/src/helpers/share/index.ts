export * from './link'
export * from './sharedAncestorRoute'
