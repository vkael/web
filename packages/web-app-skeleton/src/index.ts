import App from './App.vue'
import { GitHUbSearch } from './search/github'
import { eventBus } from 'web-pkg/src/services/eventBus'

const appInfo = {
  name: 'web-app-skeleton',
  id: 'skeleton',
  icon: 'folder',
  isFileEditor: true,
  extensions: []
}

const injectSearch = (): void => {
  eventBus.publish('app.search.register.provider', new GitHUbSearch())
}

const injectExtensions = async (api): Promise<void> => {
  // the promise is just there to showcase lazy loading of extensions
  await new Promise((resolve) => setTimeout(resolve, 2000))

  api.announceExtension({
    extension: 'txt',
    isFileEditor: true,
    newFileMenu: {
      menuTitle($gettext) {
        return $gettext('Extension from skeleton')
      }
    }
  })
}

export default {
  appInfo,
  navItems: [
    {
      name: 'skeleton',
      icon: appInfo.icon,
      route: {
        path: `/${appInfo.id}/`
      }
    }
  ],
  routes: [
    {
      name: 'skeleton',
      path: '/',
      component: App
    }
  ],
  async mounted(api) {
    await injectSearch()
    await injectExtensions(api)
  }
}
