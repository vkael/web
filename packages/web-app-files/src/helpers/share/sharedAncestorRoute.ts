import { buildShareSpaceResource, Resource, SpaceResource } from 'web-client/src/helpers'
import { basename } from 'path'
import { configurationManager } from 'web-pkg'
import { createLocationSpaces } from 'web-app-files/src/router'
import { createFileRouteOptions } from 'web-pkg/src/helpers/router'
import { RouteLocationNamedRaw } from 'vue-router'
import { AncestorMetaDataValue } from 'web-app-files/src/helpers/resource/ancestorMetaData'

export const getSharedAncestorRoute = ({
  resource,
  sharedAncestor,
  matchingSpace
}: {
  resource: Resource
  sharedAncestor: AncestorMetaDataValue
  matchingSpace: SpaceResource
}): RouteLocationNamedRaw => {
  if (resource.shareId) {
    if (resource.path === '') {
      return {}
    }
    const space = buildShareSpaceResource({
      shareId: resource.shareId,
      shareName: matchingSpace?.name || basename(resource.shareRoot),
      serverUrl: configurationManager.serverUrl
    })
    return createLocationSpaces(
      'files-spaces-generic',
      createFileRouteOptions(space, {
        path: sharedAncestor.path,
        fileId: sharedAncestor.id
      })
    )
  }
  if (!matchingSpace) {
    return {}
  }
  return createLocationSpaces(
    'files-spaces-generic',
    createFileRouteOptions(matchingSpace, {
      path: sharedAncestor.path,
      fileId: sharedAncestor.id
    })
  )
}
