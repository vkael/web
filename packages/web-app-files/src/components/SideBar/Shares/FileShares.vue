<template>
  <div id="oc-files-sharing-sidebar" class="oc-position-relative">
    <div class="oc-flex">
      <h3 v-translate class="oc-text-bold oc-text-medium oc-m-rm">Share with people</h3>
      <oc-contextual-helper
        v-if="helpersEnabled"
        class="oc-pl-xs"
        v-bind="inviteCollaboratorHelp"
      />
    </div>
    <invite-collaborator-form v-if="currentUserCanShare" key="new-collaborator" class="oc-my-s" />
    <p
      v-else
      key="no-reshare-permissions-message"
      data-testid="files-collaborators-no-reshare-permissions-message"
      v-text="noResharePermsMessage"
    />
    <div v-if="hasSharees" class="avatars-wrapper oc-flex oc-flex-middle oc-flex-between">
      <h4 class="oc-text-bold oc-my-rm" v-text="sharedWithLabel" />
    </div>
    <template v-if="hasSharees">
      <ul
        id="files-collaborators-list"
        class="oc-list oc-list-divider oc-overflow-hidden"
        :class="{ 'oc-mb-l': showSpaceMembers, 'oc-m-rm': !showSpaceMembers }"
        :aria-label="$gettext('Share receivers')"
      >
        <li v-for="collaborator in displayCollaborators" :key="collaborator.key">
          <collaborator-list-item
            :share="collaborator"
            :modifiable="isShareModifiable(collaborator)"
            :shared-parent-route="getSharedParentRoute(collaborator)"
            @on-delete="$_ocCollaborators_deleteShare_trigger"
          />
        </li>
      </ul>
      <div v-if="showShareToggle" class="oc-flex oc-flex-center">
        <oc-button
          appearance="raw"
          class="toggle-shares-list-btn"
          @click="toggleShareListCollapsed"
          v-text="collapseButtonTitle"
        />
      </div>
    </template>
    <template v-if="showSpaceMembers">
      <h4 class="oc-text-bold oc-my-s" v-text="spaceMemberLabel" />
      <ul
        id="space-collaborators-list"
        class="oc-list oc-list-divider oc-overflow-hidden oc-m-rm"
        :aria-label="spaceMemberLabel"
      >
        <li v-for="collaborator in displaySpaceMembers" :key="collaborator.key">
          <collaborator-list-item :share="collaborator" :modifiable="false" />
        </li>
      </ul>
      <div v-if="showMemberToggle" class="oc-flex oc-flex-center">
        <oc-button
          appearance="raw"
          @click="toggleMemberListCollapsed"
          v-text="collapseButtonTitle"
        />
      </div>
    </template>
  </div>
</template>

<script lang="ts">
import { mapGetters, mapActions, mapState, mapMutations } from 'vuex'
import {
  useStore,
  useCapabilityProjectSpacesEnabled,
  useCapabilityShareJailEnabled,
  useCapabilityFilesSharingResharing
} from 'web-pkg/src/composables'
import { isLocationSharesActive } from '../../../router'
import { textUtils } from '../../../helpers/textUtils'
import { ShareTypes } from 'web-client/src/helpers/share'
import InviteCollaboratorForm from './Collaborators/InviteCollaborator/InviteCollaboratorForm.vue'
import CollaboratorListItem from './Collaborators/ListItem.vue'
import {
  shareInviteCollaboratorHelp,
  shareInviteCollaboratorHelpCern
} from '../../../helpers/contextualHelpers'
import { computed, defineComponent, inject, ref, Ref, unref } from 'vue'
import { isProjectSpaceResource, Resource } from 'web-client/src/helpers'
import { getSharedAncestorRoute } from 'web-app-files/src/helpers/share'
import { AncestorMetaData } from 'web-app-files/src/helpers/resource/ancestorMetaData'
import { useShares } from 'web-app-files/src/composables'

export default defineComponent({
  name: 'FileShares',
  components: {
    InviteCollaboratorForm,
    CollaboratorListItem
  },
  setup() {
    const store = useStore()
    const sharesListCollapsed = ref(
      !store.getters.configuration.options.sidebar.shares.showAllOnLoad
    )
    const toggleShareListCollapsed = () => {
      sharesListCollapsed.value = !unref(sharesListCollapsed)
    }
    const memberListCollapsed = ref(
      !store.getters.configuration.options.sidebar.shares.showAllOnLoad
    )
    const toggleMemberListCollapsed = () => {
      memberListCollapsed.value = !unref(memberListCollapsed)
    }
    const currentUserIsMemberOfSpace = computed(() => {
      const userId = store.getters.user?.id
      if (!userId) {
        return false
      }
      return store.getters['runtime/spaces/spaceMembers'].some(
        (member) => member.collaborator?.name === userId
      )
    })

    const ancestorMetaData: Ref<AncestorMetaData> = computed(
      () => store.getters['Files/ancestorMetaData']
    )
    const getSharedAncestor = (fileId) => {
      return Object.values(unref(ancestorMetaData)).find((a) => a.id === fileId)
    }

    return {
      ...useShares(),
      resource: inject<Resource>('resource'),
      space: inject<Resource>('space'),
      sharesListCollapsed,
      toggleShareListCollapsed,
      memberListCollapsed,
      toggleMemberListCollapsed,
      currentUserIsMemberOfSpace,
      hasProjectSpaces: useCapabilityProjectSpacesEnabled(),
      hasShareJail: useCapabilityShareJailEnabled(),
      hasResharing: useCapabilityFilesSharingResharing(),
      getSharedAncestor
    }
  },
  computed: {
    ...mapGetters(['configuration']),
    ...mapGetters('runtime/spaces', ['spaceMembers']),
    ...mapState(['user']),

    inviteCollaboratorHelp() {
      const cernFeatures = !!this.configuration?.options?.cernFeatures

      if (cernFeatures) {
        const mergedHelp = shareInviteCollaboratorHelp
        mergedHelp.list = [
          ...shareInviteCollaboratorHelpCern.list,
          ...shareInviteCollaboratorHelp.list
        ]
        return mergedHelp
      }

      return shareInviteCollaboratorHelp
    },
    helpersEnabled() {
      return this.configuration?.options?.contextHelpers
    },
    sharedWithLabel() {
      return this.$gettext('Shared with')
    },
    spaceMemberLabel() {
      return this.$gettext('Space members')
    },

    collapseButtonTitle() {
      return this.sharesListCollapsed ? this.$gettext('Show more') : this.$gettext('Show less')
    },

    hasSharees() {
      return this.collaborators.length > 0
    },

    collaborators() {
      return [...this.outgoingCollaborators].sort(this.collaboratorsComparator).map((c) => {
        const collaborator = { ...c }
        collaborator.key = 'collaborator-' + collaborator.id
        if (
          collaborator.owner.name !== collaborator.fileOwner.name &&
          collaborator.owner.name !== this.user.id
        ) {
          collaborator.resharers = [collaborator.owner]
        }
        return collaborator
      })
    },

    displayCollaborators() {
      if (this.collaborators.length > 3 && this.sharesListCollapsed) {
        return this.collaborators.slice(0, 3)
      }
      return this.collaborators
    },

    displaySpaceMembers() {
      if (this.spaceMembers.length > 3 && this.memberListCollapsed) {
        return this.spaceMembers.slice(0, 3)
      }
      return this.spaceMembers
    },

    showShareToggle() {
      return this.collaborators.length > 3
    },

    showMemberToggle() {
      return this.spaceMembers.length > 3
    },

    currentUserCanShare() {
      if (this.resource.isReceivedShare() && !this.hasResharing) {
        return false
      }
      const isShareJail = this.space?.driveType === 'share'
      if (isShareJail && !this.hasResharing) {
        return false
      }

      return this.resource.canShare({ user: this.user })
    },
    noResharePermsMessage() {
      const translatedFile = this.$gettext("You don't have permission to share this file.")
      const translatedFolder = this.$gettext("You don't have permission to share this folder.")
      return this.resource.type === 'file' ? translatedFile : translatedFolder
    },
    showSpaceMembers() {
      return (
        this.space?.driveType === 'project' &&
        this.resource.type !== 'space' &&
        this.currentUserIsMemberOfSpace
      )
    },
    matchingSpace() {
      return this.space || this.spaces.find((space) => space.id === this.resource.storageId)
    }
  },
  methods: {
    ...mapActions('Files', ['deleteShare']),
    ...mapActions(['createModal', 'hideModal', 'showMessage']),
    ...mapMutations('Files', ['REMOVE_FILES']),

    collaboratorsComparator(c1, c2) {
      // Sorted by: type, direct, display name, creation date
      const c1DisplayName = c1.collaborator ? c1.collaborator.displayName : c1.displayName
      const c2DisplayName = c2.collaborator ? c2.collaborator.displayName : c2.displayName
      const name1 = c1DisplayName.toLowerCase().trim()
      const name2 = c2DisplayName.toLowerCase().trim()
      const c1UserShare = ShareTypes.containsAnyValue(ShareTypes.individuals, [c1.shareType])
      const c2UserShare = ShareTypes.containsAnyValue(ShareTypes.individuals, [c2.shareType])
      const c1DirectShare = !c1.indirect
      const c2DirectShare = !c2.indirect

      if (c1UserShare === c2UserShare) {
        if (c1DirectShare === c2DirectShare) {
          if (name1 === name2) {
            return textUtils.naturalSortCompare(c1.stime + '', c2.stime + '')
          }

          return textUtils.naturalSortCompare(name1, name2)
        }

        return c1DirectShare ? -1 : 1
      }

      return c1UserShare ? -1 : 1
    },

    $_ocCollaborators_deleteShare_trigger(share) {
      const modal = {
        variation: 'danger',
        icon: 'alarm-warning',
        title: this.$gettext('Remove share'),
        cancelText: this.$gettext('Cancel'),
        confirmText: this.$gettext('Remove'),
        message: this.$gettext('Are you sure you want to remove this share?'),
        hasInput: false,
        onCancel: this.hideModal,
        onConfirm: () => this.$_ocCollaborators_deleteShare(share)
      }

      this.createModal(modal)
    },

    async $_ocCollaborators_deleteShare(share) {
      let path = this.resource.path
      // sharing a share root from the share jail -> use resource name as path
      if (this.hasShareJail && path === '/') {
        path = `/${this.resource.name}`
      }

      const directCollaborators = this.outgoingCollaborators.filter((c) => !c.indirect)
      const lastShareId = directCollaborators.length === 1 ? directCollaborators[0].id : undefined

      try {
        await this.deleteShare({
          client: this.$client,
          share: share,
          path,
          loadIndicators: !!lastShareId
        })

        this.hideModal()
        this.showMessage({
          title: this.$gettext('Share was removed successfully')
        })
        if (lastShareId && isLocationSharesActive(this.$router, 'files-shares-with-others')) {
          this.REMOVE_FILES([{ id: lastShareId }])
        }
      } catch (error) {
        console.error(error)
        this.showMessage({
          title: this.$gettext('Failed to remove share'),
          status: 'danger'
        })
      }
    },
    getSharedParentRoute(parentShare) {
      if (!parentShare.indirect) {
        return null
      }
      const sharedAncestor = this.getSharedAncestor(parentShare.itemSource)
      if (!sharedAncestor) {
        return null
      }
      return getSharedAncestorRoute({
        resource: this.resource,
        sharedAncestor,
        matchingSpace: this.matchingSpace
      })
    },

    // fixMe: head-breaking logic
    isShareModifiable(collaborator) {
      const isPersonalSpaceShare = !isProjectSpaceResource(this.space)
      const isPersonalMember = this.currentUserIsMemberOfSpace
      const isIndirectPersonalCollaborator = collaborator.indirect
      const isProjectSpaceShare = !isPersonalSpaceShare
      const isManager = this.space?.isManager(this.user)

      if (isPersonalSpaceShare && isPersonalMember && isManager) {
        return true
      }

      if (isPersonalSpaceShare && !isIndirectPersonalCollaborator) {
        return true
      }

      if (isProjectSpaceShare && isManager) {
        return true
      }

      return false
    }
  }
})
</script>

<style>
.avatars-wrapper {
  height: 40px;
}
</style>
