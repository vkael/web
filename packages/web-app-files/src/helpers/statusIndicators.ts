import { ShareTypes } from 'web-client/src/helpers/share'
import { eventBus } from 'web-pkg'
import { SideBarEventTopics } from 'web-pkg/src/composables/sideBar'
import { createLocationShares } from 'web-app-files/src/router'
import { Resource } from 'web-client'
import { AncestorMetaData } from 'web-app-files/src/helpers/resource/ancestorMetaData'

// dummy to trick gettext string extraction into recognizing strings
const $gettext = (str) => {
  return str
}

const isUserShare = (shareTypes) => {
  return ShareTypes.containsAnyValue(ShareTypes.authenticated, shareTypes ?? [])
}

const isLinkShare = (shareTypes) => {
  return ShareTypes.containsAnyValue(ShareTypes.unauthenticated, shareTypes ?? [])
}

const shareUserIconDescribedBy = ({ isDirect }) => {
  return isDirect
    ? $gettext('This item is directly shared with others.')
    : $gettext('This item is shared with others through one of the parent folders.')
}

const shareLinkDescribedBy = ({ isDirect }) => {
  return isDirect
    ? $gettext('This item is directly shared via links.')
    : $gettext('This item is shared via links through one of the parent folders.')
}

const getUserIndicator = ({ resource, isDirect, isIncoming = false }) => {
  return {
    id: `files-sharing-${resource.getDomSelector()}`,
    accessibleDescription: shareUserIconDescribedBy({ isDirect }),
    label: isIncoming ? $gettext('Shared with you') : $gettext('Show invited people'),
    icon: 'group',
    target: 'sharing',
    type: isDirect ? 'user-direct' : 'user-indirect',
    handler: (resource, panel, $router) => {
      if (isIncoming) {
        $router.push(createLocationShares('files-shares-with-me'))
        return
      }
      eventBus.publish(SideBarEventTopics.openWithPanel, `${panel}#peopleShares`)
    }
  }
}

const getLinkIndicator = ({ resource, isDirect }) => {
  return {
    id: `file-link-${resource.getDomSelector()}`,
    accessibleDescription: shareLinkDescribedBy({ isDirect }),
    label: $gettext('Show links'),
    icon: 'link',
    target: 'sharing',
    type: isDirect ? 'link-direct' : 'link-indirect',
    handler: (resource, panel) => {
      eventBus.publish(SideBarEventTopics.openWithPanel, `${panel}#linkShares`)
    }
  }
}

export const getIndicators = ({
  resource,
  ancestorMetaData
}: {
  resource: Resource
  ancestorMetaData: AncestorMetaData
}) => {
  const indicators = []
  const parentShareTypes = Object.values(ancestorMetaData).reduce((acc: any, data: any) => {
    acc.push(...(data.shareTypes || []))
    return acc
  }, [])
  const isDirectUserShare = isUserShare(resource.shareTypes)
  if (isDirectUserShare || isUserShare(parentShareTypes)) {
    indicators.push(getUserIndicator({ resource, isDirect: isDirectUserShare }))
  } else if (resource.isReceivedShare()) {
    indicators.push(getUserIndicator({ resource, isDirect: false, isIncoming: true }))
  }
  const isDirectLinkShare = isLinkShare(resource.shareTypes)
  if (isDirectLinkShare || isLinkShare(parentShareTypes)) {
    indicators.push(getLinkIndicator({ resource, isDirect: isDirectLinkShare }))
  }
  return indicators
}
