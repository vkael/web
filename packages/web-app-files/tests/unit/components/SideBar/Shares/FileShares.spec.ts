import FileShares from 'web-app-files/src/components/SideBar/Shares/FileShares.vue'
import { mock } from 'jest-mock-extended'
import { Resource } from 'web-client'
import { SpaceResource } from 'web-client/src/helpers'
import { v4 as uuidV4 } from 'uuid'
import { Share } from 'web-client/src/helpers/share'
import {
  createStore,
  defaultPlugins,
  mount,
  shallowMount,
  defaultStoreMockOptions,
  defaultComponentMocks,
  defaultStubs,
  RouteLocation
} from 'web-test-helpers'

const getCollaborator = () => ({
  shareType: 0,
  id: uuidV4(),
  collaborator: {
    name: 'einstein',
    displayName: 'Albert Einstein',
    additionalInfo: 'einstein@example.org'
  },
  owner: {
    name: 'admin',
    displayName: 'Admin',
    additionalInfo: 'admin@example.org'
  },
  fileOwner: {
    name: 'admin',
    displayName: 'Admin',
    additionalInfo: 'admin@example.org'
  },
  file: {},
  stime: '1639152810',
  permissions: 15,
  itemSource: '1',
  path: "/Neuer Ordner-'singe'",
  key: 'collaborator-f5c28709-b921-4ec8-b39a-4c243709b514'
})

describe('FileShares', () => {
  describe('invite collaborator form', () => {
    it('renders the form when the resource can be shared', () => {
      const resource = mock<Resource>({ isReceivedShare: () => false, canShare: () => true })
      const { wrapper } = getWrapper({ resource })
      expect(wrapper.find('invite-collaborator-form-stub').exists()).toBeTruthy()
    })
    it('does not render the form when the resource can not be shared', () => {
      const resource = mock<Resource>({ isReceivedShare: () => false, canShare: () => false })
      const { wrapper } = getWrapper({ resource })
      expect(wrapper.find('invite-collaborator-form-stub').exists()).toBeFalsy()
    })
    it('does render the form when the resource is a received share and re-sharing is enabled', () => {
      const resource = mock<Resource>({ isReceivedShare: () => true, canShare: () => true })
      const { wrapper } = getWrapper({ resource })
      expect(wrapper.find('invite-collaborator-form-stub').exists()).toBeTruthy()
    })
    it('does not render the form when the resource is a received share and re-sharing is disabled', () => {
      const resource = mock<Resource>({ isReceivedShare: () => true, canShare: () => true })
      const { wrapper } = getWrapper({ resource, hasReSharing: false })
      expect(wrapper.find('invite-collaborator-form-stub').exists()).toBeFalsy()
    })
  })

  describe('collaborators list', () => {
    let collaborators
    beforeEach(() => {
      collaborators = [getCollaborator(), getCollaborator(), getCollaborator(), getCollaborator()]
    })

    it('renders sharedWithLabel and sharee list', () => {
      const { wrapper } = getWrapper({ collaborators })
      expect(wrapper.find('#files-collaborators-list').exists()).toBeTruthy()
      expect(wrapper.findAll('#files-collaborators-list li').length).toBe(collaborators.length)
      expect(wrapper.html()).toMatchSnapshot()
    })
    it('reacts on delete events', async () => {
      const spyOnCollaboratorDeleteTrigger = jest
        .spyOn((FileShares as any).methods, '$_ocCollaborators_deleteShare_trigger')
        .mockImplementation()
      const { wrapper } = getWrapper({ collaborators })
      ;(wrapper.findComponent<any>('collaborator-list-item-stub').vm as any).$emit('onDelete')
      await wrapper.vm.$nextTick()
      expect(spyOnCollaboratorDeleteTrigger).toHaveBeenCalledTimes(1)
    })
    it('correctly passes the shared parent route to the collaborator list item for indirect shares', () => {
      const indirectCollaborator = { ...getCollaborator(), indirect: true }
      const ancestorMetaData = {
        '/somePath': { id: indirectCollaborator.itemSource }
      }
      const { wrapper } = getWrapper({ collaborators: [indirectCollaborator], ancestorMetaData })
      const listItemStub = wrapper.findComponent<any>('collaborator-list-item-stub')
      expect(listItemStub.props('sharedParentRoute')).toBeTruthy()
      expect(listItemStub.props('modifiable')).toBeFalsy()
    })
    it('toggles the share list', async () => {
      const showAllOnLoad = true
      const { wrapper } = getWrapper({ mountType: mount, collaborators })
      expect(wrapper.vm.sharesListCollapsed).toBe(!showAllOnLoad)
      await wrapper.find('.toggle-shares-list-btn').trigger('click')
      expect(wrapper.vm.sharesListCollapsed).toBe(showAllOnLoad)
    })
    it('share should be modifiable if its personal space share', async () => {
      const space = mock<SpaceResource>({ driveType: 'personal' })
      const { wrapper } = getWrapper({ space, mountType: shallowMount, collaborators })
      expect(wrapper.vm.isShareModifiable(collaborators[0])).toBe(true)
    })
    it('share should not be modifiable if its not personal space share', async () => {
      const space = mock<SpaceResource>({ driveType: 'project' })
      const { wrapper } = getWrapper({ space, mountType: shallowMount, collaborators })
      expect(wrapper.vm.isShareModifiable(collaborators[0])).toBe(false)
    })
    it('share should not be modifiable if collaborator is indirect', async () => {
      const space = mock<SpaceResource>({ driveType: 'personal' })
      const { wrapper } = getWrapper({ space, mountType: shallowMount, collaborators })
      collaborators[0]['indirect'] = true
      expect(wrapper.vm.isShareModifiable(collaborators[0])).toBe(false)
    })
    it('share should not be modifiable if user is not manager', async () => {
      const space = mock<SpaceResource>({ driveType: 'personal' }) as any
      ;(space as any).isManager = jest.fn(() => false)
      collaborators[0]['indirect'] = true
      const { wrapper } = getWrapper({ space, mountType: shallowMount, collaborators })
      expect(wrapper.vm.isShareModifiable(collaborators[0])).toBe(false)
    })
  })

  describe('current space', () => {
    it('loads space members if a space is given and the current user is member', () => {
      const user = { id: '1' }
      const space = mock<SpaceResource>({ driveType: 'project' })
      const spaceMembers = [{ collaborator: { name: user.id } }, { collaborator: { name: 2 } }]
      const collaborator = getCollaborator()
      collaborator.collaborator = { ...collaborator.collaborator, name: user.id }
      const { wrapper } = getWrapper({ space, collaborators: [collaborator], user, spaceMembers })
      expect(wrapper.find('#files-collaborators-list').exists()).toBeTruthy()
      expect(wrapper.findAll('#files-collaborators-list li').length).toBe(1)
      expect(wrapper.html()).toMatchSnapshot()
    })
    it('does not load space members if a space is given but the current user not a member', () => {
      const user = { id: '1' }
      const space = mock<SpaceResource>({ driveType: 'project' })
      const spaceMembers = [{ collaborator: { name: `${user}-2` } }]
      const collaborator = getCollaborator()
      collaborator.collaborator = { ...collaborator.collaborator, name: user.id }
      const { wrapper } = getWrapper({ space, collaborators: [collaborator], user, spaceMembers })
      expect(wrapper.find('#space-collaborators-list').exists()).toBeFalsy()
    })
  })

  describe('"$_ocCollaborators_deleteShare" method', () => {
    it('calls "deleteShare" when successful', async () => {
      const { wrapper } = getWrapper()
      const deleteShareSpy = jest.spyOn(wrapper.vm, 'deleteShare')
      const share = mock<Share>()
      await wrapper.vm.$_ocCollaborators_deleteShare(share)
      expect(deleteShareSpy).toHaveBeenCalled()
    })
    it('shows a message when an error occurs', async () => {
      jest.spyOn(console, 'error').mockImplementation(() => undefined)
      const { wrapper } = getWrapper()
      jest.spyOn(wrapper.vm, 'deleteShare').mockRejectedValue(new Error())
      const showMessageSpy = jest.spyOn(wrapper.vm, 'showMessage')
      const share = mock<Share>()
      await wrapper.vm.$_ocCollaborators_deleteShare(share)
      expect(showMessageSpy).toHaveBeenCalled()
    })
    it('removes file when the last share on the "Shared with others"-page has been removed', async () => {
      const { wrapper } = getWrapper({
        collaborators: [getCollaborator()],
        currentRouteName: 'files-shares-with-others'
      })
      const deleteShareSpy = jest.spyOn(wrapper.vm, 'deleteShare')
      const removeFilesSpy = jest.spyOn(wrapper.vm, 'REMOVE_FILES')
      const share = mock<Share>()
      await wrapper.vm.$_ocCollaborators_deleteShare(share)
      expect(deleteShareSpy).toHaveBeenCalled()
      expect(removeFilesSpy).toHaveBeenCalled()
    })
  })
})

function getWrapper({
  mountType = shallowMount,
  resource = mock<Resource>({ isReceivedShare: () => false, canShare: () => true }),
  hasReSharing = true,
  space = mock<SpaceResource>(),
  collaborators = [],
  spaceMembers = [],
  user = { id: '1' },
  showAllOnLoad = true,
  currentRouteName = 'files-spaces-generic',
  ancestorMetaData = {}
} = {}) {
  const storeOptions = {
    ...defaultStoreMockOptions,
    state: { user },
    getters: {
      ...defaultStoreMockOptions.getters,
      user: () => user,
      capabilities: jest.fn(() => ({
        files_sharing: { resharing: hasReSharing }
      })),
      configuration: jest.fn(() => ({
        options: { contextHelpers: true, sidebar: { shares: { showAllOnLoad } } }
      }))
    }
  }
  storeOptions.modules.Files.getters.ancestorMetaData.mockReturnValue(ancestorMetaData)
  storeOptions.modules.runtime.modules.spaces.getters.spaceMembers.mockImplementation(
    () => spaceMembers
  )
  storeOptions.modules.Files.getters.outgoingCollaborators.mockReturnValue(collaborators)
  const store = createStore(storeOptions)
  return {
    wrapper: mountType(FileShares, {
      global: {
        plugins: [...defaultPlugins(), store],
        mocks: defaultComponentMocks({
          currentRoute: mock<RouteLocation>({ name: currentRouteName })
        }),
        provide: {
          resource,
          space,
          incomingParentShare: {}
        },
        stubs: {
          ...defaultStubs,
          OcButton: false,
          'invite-collaborator-form': true,
          'collaborator-list-item': true
        }
      }
    })
  }
}
