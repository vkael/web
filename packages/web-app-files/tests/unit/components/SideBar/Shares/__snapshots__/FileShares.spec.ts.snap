// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileShares collaborators list renders sharedWithLabel and sharee list 1`] = `
<div class="oc-position-relative" id="oc-files-sharing-sidebar">
  <div class="oc-flex">
    <h3 class="oc-text-bold oc-text-medium oc-m-rm" data-current-language="en" data-msgid="Share with people">Share with people</h3>
    <oc-contextual-helper-stub class="oc-pl-xs" endtext="" list="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" readmorelink="https://doc.owncloud.com/webui/next/owncloud_web/web_for_users.html#sharing" text="Enter a name or group to share this item." title="Invite specific people"></oc-contextual-helper-stub>
  </div>
  <invite-collaborator-form-stub class="oc-my-s" invitelabel="" savebuttonlabel="Share"></invite-collaborator-form-stub>
  <div class="avatars-wrapper oc-flex oc-flex-middle oc-flex-between">
    <h4 class="oc-text-bold oc-my-rm">Shared with</h4>
  </div>
  <ul aria-label="Share receivers" class="oc-list oc-list-divider oc-overflow-hidden oc-m-rm" id="files-collaborators-list">
    <li>
      <collaborator-list-item-stub modifiable="true" share="[object Object]"></collaborator-list-item-stub>
    </li>
    <li>
      <collaborator-list-item-stub modifiable="true" share="[object Object]"></collaborator-list-item-stub>
    </li>
    <li>
      <collaborator-list-item-stub modifiable="true" share="[object Object]"></collaborator-list-item-stub>
    </li>
    <li>
      <collaborator-list-item-stub modifiable="true" share="[object Object]"></collaborator-list-item-stub>
    </li>
  </ul>
  <div class="oc-flex oc-flex-center">
    <button class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw toggle-shares-list-btn" type="button">Show less</button>
  </div>
  <!--v-if-->
</div>
`;

exports[`FileShares current space loads space members if a space is given and the current user is member 1`] = `
<div class="oc-position-relative" id="oc-files-sharing-sidebar">
  <div class="oc-flex">
    <h3 class="oc-text-bold oc-text-medium oc-m-rm" data-current-language="en" data-msgid="Share with people">Share with people</h3>
    <oc-contextual-helper-stub class="oc-pl-xs" endtext="" list="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" readmorelink="https://doc.owncloud.com/webui/next/owncloud_web/web_for_users.html#sharing" text="Enter a name or group to share this item." title="Invite specific people"></oc-contextual-helper-stub>
  </div>
  <invite-collaborator-form-stub class="oc-my-s" invitelabel="" savebuttonlabel="Share"></invite-collaborator-form-stub>
  <div class="avatars-wrapper oc-flex oc-flex-middle oc-flex-between">
    <h4 class="oc-text-bold oc-my-rm">Shared with</h4>
  </div>
  <ul aria-label="Share receivers" class="oc-list oc-list-divider oc-overflow-hidden oc-mb-l" id="files-collaborators-list">
    <li>
      <collaborator-list-item-stub modifiable="false" share="[object Object]"></collaborator-list-item-stub>
    </li>
  </ul>
  <!--v-if-->
  <h4 class="oc-text-bold oc-my-s">Space members</h4>
  <ul aria-label="Space members" class="oc-list oc-list-divider oc-overflow-hidden oc-m-rm" id="space-collaborators-list">
    <li>
      <collaborator-list-item-stub modifiable="false" share="[object Object]"></collaborator-list-item-stub>
    </li>
    <li>
      <collaborator-list-item-stub modifiable="false" share="[object Object]"></collaborator-list-item-stub>
    </li>
  </ul>
  <!--v-if-->
</div>
`;
